### DISCUSSION TITLE
Print Layout design and printing function for Native cross platform desktop application

### DISCUSSION THREAD

**Discussion by praktanss** [AvaloniaUI/Avalonia#14820] [Ideas]
Similar to RDLC what is inbuilt provision for cross platform native application reporting with print layout design and printing. Also end user should be capable to create and design layout of report format and set to view document for printing. It is very important and integral part of project to deliver business solutions.

**Comment by timunie** [comment]
I believe printing is not on the list in the near future. #12567

**Comment by thevortexcloud** [comment]
It sounds like you want a PDF. There are already many libraries that can generate them, some are free and some are paid. 

EG QuestPDF and Telerik Reports. Telerik Reports specifically ships with its own report designer.

Either way as long as you can convert the PDF/report output into a bitmap, it's fairly trivial to make a previewer in Avalonia. I have made one myself before using QuestPDF and it worked pretty well.

There are also many ways to print a PDF. EG on Windows you can run Adobe Acrobat in the background and on Linux I think you can just send it direct to CUPS.

In any case I would say this is well outside the scope of a UI project.

**Reply by praktanss** [reply]
Thanks a Million for prompt advise.
We are looking for inbuilt report designer like RDLC which can preview, Print, Direct print, Export to MS Document and PDF etc.. User can design Layout of the Order or Invoice document and set for Transaction, also for MIS Tabular reports with Cross Tab & Matrix format etc.. We prefer inbuilt tool to avoid dependency and version conflict when upgrade to next SDK version of the application, that time it should be compatible without any workaround. 
It would be better if Avalonia consider this on top priority because complex enterprise applications like ERP need this feature to select Avalonia for product development. I am also thinking to migrate our ERP to Avalonia UI. 
Advise is welcome to take decision to go with it.

**Reply by MikeCodesDotNET** [reply]
The reporting features you've suggested are hugely complex and unlikely to be included in the project. 

However, if you're interested in sponsoring the development of these features, we'd be happy to discuss this further.

**Reply by thevortexcloud** [reply]
There is nothing stopping you from using something like QuestPDF and Avalonia to make your own designer. You typically write templates in C# for it, but you could make some other format which gets parsed into C# code (and then fed into a source generator)/into a QuestPDF document. They have a fairly simple example on how to make an invoice.

https://www.questpdf.com/getting-started.html

> It would be better if Avalonia consider this on top priority because complex enterprise applications 

I am still not clear why you think this should be Avalonia's responsibility though. Avalonia is primarily a UI framework. I am pretty sure WPF can't do this either out of the box and that was made by Microsoft with a significantly bigger budget than Avalonia which only has a full time team of a dozen or so people. If you really want Avalonia to do this for you, I woud suggest contacting their sales team and potentially paying Avalonia OU to do a custom development for you. 

One thing that may work in future is Hybrid XPF. Which will let you run WPF code inside Avalonia cross platform. Which means you would be able to use existing WPF designers like ones from Dev Express. Which there is apparently an active fix being worked on for.

https://avaloniaui.net/XPF/Packages/devexpress

EDIT:  If you do end up sponsoring this as Mike suggested, I would personally would prefer something like this was made open source/made available as an official Avalonia package. But of course that would be up to what ever agreement you make with the devs.

**Reply by praktanss** [reply]
Noted. Thanks!
Kindly advise sponsoring process and cost estimation policy for my reference then it would be better to take this further. if we select AvaloniaUI as our product Base Framework then we may required many things to design, Development and integrate into the ERP Application. I would like to be associated with you all as POC is undergoing right now for evolution for framework selection. Also advise the plus points to recommend Avalonia for business application.

**Reply by MikeCodesDotNET** [reply]
We can only give you an accurate cost with a specification and some time to investigate. That said, such a project would likely be a six-figure project, requiring several months to complete.

**Reply by praktanss** [reply]
Thanks for advise. We will back to you if we select Avalonia for our product.

**Comment by timunie** [comment]
Maybe this will help in the future: https://www.fast-report.com/en/news/show/announcement-fastreport-avalonia/

**Reply by praktanss** [reply]
Thanks for your advise.
